test_data/asm/expand_fixed.asm

```diff
@@ -0,0 +1,141 @@
+let compute_length_of_repeated_part: ArrayTerm[], (-> int) -> Result<int, string> = |terms, degree| {
+    let (_, res) = std::array::fold(terms, (false, Result::Ok(0)), |(found_repeated, l), term| {
+        match l {
+            Result::Err(e) => (found_repeated, Result::Err(e)),
+            Result::Ok(len) => match term {
+                // we can have `[]*` as many times as we want
+                ArrayTerm::Repeat([]) => (found_repeated, Result::Ok(len)),
+                ArrayTerm::Repeat(a) => {
+                    if found_repeated {
+                        (true, Result::Err(MORE_THAN_ONE_REPEATED_ERROR))
+                    } else {
+                        (true, Result::Ok(len))
+                    }
+                },
+                ArrayTerm::Once(a) => (found_repeated, Result::Ok(len + std::array::len(a)))
+            }
+        }
+    });
+    match res {
+        Result::Ok(total_size_of_non_repeated) => if total_size_of_non_repeated <= degree() {
+            Result::Ok(degree() - total_size_of_non_repeated)
+        } else {
+            Result::Err(NON_REPEATED_SIZE_EXCEEDS_DEGREE_ERROR)
+        },
+        Result::Err(e) => Result::Err(e)
+    }
+};
+
+enum OffsetSearch {
+    Offset(int),
+    Value(int),
+}
+
+// returns a function representing the array expression
+let expand: ArrayTerm[], (-> int) -> Result<(int -> int), string> = |terms, degree| {
+    // return early if all terms are constant with the same value
+    let (constant_value, found_repeated_item) = std::array::fold(terms, (Result::Ok(Option::None), Result::Ok(false)), |(r, found), term| {
+        let (a, is_repeated) = match term {
+            ArrayTerm::Once(a) => (a, false),
+            ArrayTerm::Repeat(a) => (a, std::array::len(a) > 0),
+        };
+
+        (match r {
+            Result::Ok(c) => {
+                std::array::fold(a, Result::Ok(c), |e, value| {
+                    match e {
+                        Result::Ok(Option::None) => Result::Ok(Option::Some(value)),
+                        Result::Ok(Option::Some(prev_value)) => if value == prev_value { Result::Ok(Option::Some(value)) } else { Result::Err(()) },
+                        Result::Err(()) => Result::Err(())
+                    }
+                })
+            },
+            Result::Err(()) => Result::Err(())
+        }, match found {
+            Result::Ok(v) => if v && is_repeated { Result::Err(MORE_THAN_ONE_REPEATED_ERROR) } else { Result::Ok(v || is_repeated) },
+            Result::Err(e) => Result::Err(e),
+        })
+    });
+
+    match (constant_value, found_repeated_item) {
+        // we found more than one repeated term, error out
+        (_, Result::Err(s)) => Result::Err(s),
+        // This is not strict enough and would allow expressions whose size does not match the degree
+        // We should fix this by ensuring that:
+        // - the minimum size of the expanded expression is smaller than the minimum degree of the machine
```

Comment:
what happens when the expansion cannot get to the degree?
e.g.: `[1,2]* + [3]`
this would always lead to an "odd" sized expansion no?